File: rego-contract/pdo/rego/plugins/rego_token.py
# Copyright 2023 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import base64
import json
import logging

from pdo.contract import invocation_request


import pdo.client.builder as pbuilder
import pdo.client.builder.command as pcommand
import pdo.client.builder.contract as pcontract
import pdo.client.builder.shell as pshell
import pdo.client.commands.contract as pcontract_cmd

import pdo.exchange.plugins.token_object as token_object
import pdo.identity.plugins.policy_agent as policy_agent
from pdo.contracts.guardian.common.guardian_service import GuardianServiceClient

__all__ = [
    "op_initialize",
    "op_get_verifying_key",
    "op_get_contract_metadata",
    "op_get_contract_code_metadata",
    "op_get_asset_type_identifier",
    "op_get_issuer_authority",
    "op_get_authority",
    "op_transfer",
    "op_escrow",
    "op_release",
    "op_claim",
    "op_do_download",
    "cmd_mint_tokens",
    "cmd_transfer_assets",
    "cmd_do_download",
    "do_rego_token",
    "do_rego_token_contract",
    "load_commands",
]

## -----------------------------------------------------------------
## inherited operations
## -----------------------------------------------------------------
op_get_verifying_key = token_object.op_get_verifying_key
op_get_contract_metadata = token_object.op_get_contract_metadata
op_get_contract_code_metadata = token_object.op_get_contract_code_metadata
op_get_asset_type_identifier = token_object.op_get_asset_type_identifier
op_get_issuer_authority = token_object.op_get_issuer_authority
op_get_authority = token_object.op_get_authority
op_transfer = token_object.op_transfer
op_escrow = token_object.op_escrow
op_release = token_object.op_release
op_claim = token_object.op_claim
op_initialize = token_object.op_initialize
op_register_trusted_issuer = policy_agent.op_register_trusted_issuer
op_list_trusted_issuers = policy_agent.op_list_trusted_issuers

cmd_mint_tokens = token_object.cmd_mint_tokens
cmd_transfer_assets = token_object.cmd_transfer_assets
cmd_register_trusted_issuer = policy_agent.cmd_register_trusted_issuer
cmd_list_trusted_issuers = policy_agent.cmd_list_trusted_issuers

logger = logging.getLogger(__name__)


## -----------------------------------------------------------------
## -----------------------------------------------------------------
class op_do_download(pcontract.contract_op_base):
    """op_download implements the full end-to-end download of data gated on a
    policy decision credential issued by the rego_policy_agent.
    """

    name = "op_download"
    help = "download using a rego policy decision credential"

    @classmethod
    def add_arguments(cls, subparser):
        subparser.add_argument(
            "--vc",
            help="Input policy decision vc for the token",
            type=pbuilder.invocation_parameter,
            required=True,
        )

        subparser.add_argument(
            "-u", "--url", help="URL for the guardian service", type=str, required=True
        )

    @classmethod
    def invoke(cls, state, session_params, vc, url, **kwargs):
        session_params["commit"] = False

        # send the request to the contract to create a capability for the guardian
        params = {}
        params["policy_vc"] = vc

        message = invocation_request("do_download", **params)
        capability = pcontract_cmd.send_to_contract(state, message, **session_params)

        capability = json.loads(capability)

        cls.log_invocation(message, capability)

        # process the capability that was created
        service_client = GuardianServiceClient(url)

        # send the capability to the guardian, this returns a dictionary
        result = service_client.process_capability(**capability)
        return base64.b64decode(result)


## -----------------------------------------------------------------
## -----------------------------------------------------------------
class cmd_do_download(pcommand.contract_command_base):
    """cmd_download implements the full end-to-end download of data gated on a
    policy decision credential issued by the rego_policy_agent.
    """

    name = "do_download"
    help = "download using a rego policy decision credential"

    @classmethod
    def add_arguments(cls, subparser):
        subparser.add_argument(
            "--vc-file",
            help="Input policy decision vc file for the token",
            type=str,
            required=True,
        )
        subparser.add_argument(
            "--output-file",
            help="where to save the downloaded data",
            type=str,
            required=True,
        )
        subparser.add_argument(
            "-u", "--url", help="URL for the guardian service", type=str
        )

    @classmethod
    def invoke(cls, state, context, vc_file, output_file, url=None, **kwargs):
        save_file = pcontract_cmd.get_contract_from_context(state, context)
        if not save_file:
            raise ValueError("token has not been created")

        if url is None:
            guardian_context = context.get_context("data_guardian_context")
            url = guardian_context["url"]

        with open(vc_file, "r") as fp:
            signed_credential_data = json.load(fp)

        session = pbuilder.SessionParameters(save_file=save_file)
        result = pcontract.invoke_contract_op(
            op_do_download,
            state,
            context,
            session,
            vc=signed_credential_data,
            url=url,
            **kwargs,
        )
        with open(output_file, "wb") as fp:
            fp.write(result)

        return result


## -----------------------------------------------------------------
## Create the generic, shell independent version of the aggregate command
## -----------------------------------------------------------------
__operations__ = [
    op_initialize,
    op_get_verifying_key,
    op_get_contract_metadata,
    op_get_contract_code_metadata,
    op_get_asset_type_identifier,
    op_get_issuer_authority,
    op_get_authority,
    op_transfer,
    op_escrow,
    op_release,
    op_claim,
    op_do_download,
    op_register_trusted_issuer,
    op_list_trusted_issuers,
]

do_rego_token_contract = pcontract.create_shell_command(
    "rego_token_contract", __operations__
)

__commands__ = [
    cmd_mint_tokens,
    cmd_transfer_assets,
    cmd_do_download,
    cmd_register_trusted_issuer,
    cmd_list_trusted_issuers,
]

do_rego_token = pcommand.create_shell_command("rego_token", __commands__)


## -----------------------------------------------------------------
## Enable binding of the shell independent version to a pdo-shell command
## -----------------------------------------------------------------
def load_commands(cmdclass):
    pshell.bind_shell_command(cmdclass, "rego_token", do_rego_token)
    pshell.bind_shell_command(
        cmdclass, "rego_token_contract", do_rego_token_contract
    )

File: rego-contract/src/methods/rego_token.cpp
/* Copyright 2026 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <stdlib.h>
#include <stdint.h>
#include <string>

#include "Dispatch.h"

#include "Cryptography.h"
#include "KeyValue.h"
#include "Environment.h"
#include "Message.h"
#include "Response.h"
#include "Types.h"
#include "Util.h"
#include "Value.h"
#include "WasmExtensions.h"

#include "contract/base.h"
#include "exchange/token_object.h"
#include "identity/policy_agent.h"
#include "rego/rego_policy_agent.h" // REGO_OPERATION_SCHEMA
#include "rego/rego_token.h"

bool ww::rego::rego_token::do_operation(
    const Message &msg,
    const Environment &env,
    Response &rsp)
{
    ASSERT_INITIALIZED(rsp);

    ASSERT_SUCCESS(rsp, msg.validate_schema(REGO_TOKEN_PARAM_SCHEMA),
                   "invalid request, missing required parameters");

    ww::value::Object vc_object_in;
    ASSERT_SUCCESS(rsp, msg.get_value("policy_vc", vc_object_in), "missing required parameter; policy_vc");

    // the rego_policy_agent issues "policy_decision" credentials; verify the
    // presented credential against the issuer registered for that type
    ww::identity::VerifiableCredential vc_in;
    ASSERT_SUCCESS(rsp, ww::identity::policy_agent::verify_credential(vc_object_in, vc_in, "policy_decision"), "invalid request, ill-formed credential");

    // the credential's claims are the merged operation the policy authorized:
    // { "name": <operation>, "parameters": { ... } }. The policy -- not the token --
    // decides which operation to run and with what parameters.
    ww::value::Object &operation = vc_in.credential_.credentialSubject_.claims_;
    ASSERT_SUCCESS(rsp, operation.validate_schema(REGO_OPERATION_SCHEMA),
                   "invalid request, ill-formed operation in credential");

    const char *operation_name = operation.get_string("name");
    ASSERT_SUCCESS(rsp, operation_name != nullptr, "invalid request, operation missing name");

    ww::value::Object parameters;
    ASSERT_SUCCESS(rsp, operation.get_value("parameters", parameters),
                   "invalid request, operation missing parameters");

    // build the guardian capability for the named operation with its parameters
    ww::value::Object result;
    ASSERT_SUCCESS(rsp, ww::exchange::token_object::create_operation_package(
                            operation_name, parameters, result),
                   "unexpected error: failed to generate capability");

    // this assumes that generating the capability does not change state, depending on
    // how the nonce is created this may need to change.
    return rsp.value(result, false);
}
